readthedocs/oauth/services/gitlab.py

```diff
@@ -0,0 +1,330 @@
+        """
+        Get webhook JSON data to post to the API.
+
+        See: http://doc.gitlab.com/ce/api/projects.html#add-project-hook
+        """
+        return json.dumps({
+            'id': repo_id,
+            'push_events': True,
+            'tag_push_events': True,
+            'url': 'https://{domain}{path}'.format(
+                domain=settings.PRODUCTION_DOMAIN,
+                path=reverse(
+                    'api_webhook',
+                    kwargs={
+                        'project_slug': project.slug,
+                        'integration_pk': integration.pk,
+                    },
+                ),
+            ),
+
+            # Optional
+            'issues_events': False,
+            'merge_requests_events': False,
+            'note_events': False,
+            'job_events': False,
+            'pipeline_events': False,
+            'wiki_events': False,
+        })
+
+    def setup_webhook(self, project):
+        """
+        Set up GitLab project webhook for project.
+
+        :param project: project to set up webhook for
+        :type project: Project
+        :returns: boolean based on webhook set up success
+        :rtype: bool
+        """
+        session = self.get_session()
+        owner, repo = build_utils.get_gitlab_username_repo(url=project.repo)
+        integration, _ = Integration.objects.get_or_create(
+            project=project,
+            integration_type=Integration.GITLAB_WEBHOOK,
+        )
+
+        # The ID or URL-encoded path of the project
+        # https://docs.gitlab.com/ce/api/README.html#namespaced-path-encoding
+        repo_id = '{}%2F{}'.format(owner, repo)
+
+        data = self.get_webhook_data(repo_id, integration, project)
+        resp = None
+        try:
+            resp = session.post(
+                '{url}/api/v4/projects/{repo_id}/hooks'.format(
+                    url=self.adapter.provider_base_url,
+                    repo_id=repo_id,
+                ),
+                data=data,
+                headers={'content-type': 'application/json'},
+            )
+            if resp.status_code == 201:
+                integration.provider_data = resp.json()
+                integration.save()
+                log.info(
+                    'GitLab webhook creation successful for project: %s',
+                    project,
+                )
+                return (True, resp)
+        except (RequestException, ValueError):
+            log.exception(
+                'GitLab webhook creation failed for project: %s',
+                project,
+            )
+        else:
+            log.exception(
+                'GitLab webhook creation failed for project: %s',
+                project,
+            )
+            return (False, resp)
+
+    def update_webhook(self, project, integration):
+        # TODO: to implement
+        pass
+
+    # @classmethod
+    # def get_token_for_project(cls, project, force_local=False):
```

Comment:
I think this is only needed by Github, but I don't understand how it works yet...